spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -1,51 +1,266 @@
+
+* [**ApplySnapshotChunk:**](./abci++_methods_002_draft.md#applysnapshotchunk) Used by Tendermint to hand snapshot chunks to the Application.
+
+### Other methods
+
+Additionally, there is a [**Flush**](./abci++_methods_002_draft.md#flush) method that is called on every connection,
+and an [**Echo**](./abci++_methods_002_draft.md#echo) method that is just for debugging.
 
 >**TODO** Figure out what to do with this.
 
 More details on managing state across connections can be found in the section on
 [ABCI Applications](../abci/apps.md).
 
+## Next-block execution vs. same-block execution
+[&uparrow; Back to Outline](#outline)
+
+With ABCI++ predecessor, ABCI, the only moment when the Application had access to a
+block was when it was decided. This led to a block execution model, called _next-block
+execution_, where some fields hashed in a block header refer to the execution of the
+previous block, namely:
+
+* the merkle root of the Application's state
+* the transaction results
+* the consensus parameter updates
+* the validator updates
+
+With ABCI++, an Application may decide to keep using the next-block execution model;
+however the new methods introduced, `PrepareProposal` and `ProcessProposal` allow
+for a new execution model, called _same-block execution_. An Application implementing
+this execution model, upon receiving a raw proposal via `RequestPrepareProposal`
+and potentially modifying its transaction list,
+fully executes the resulting prepared proposal as though it was the decided block.
+The results of the block execution are used as follows:
+
+* the Application keeps the events generated and provides them if `FinalizeBlock`
+  is finally called on this prepared proposal.
+* the merkle root resulting from executing the prepared proposal is provided in
+  `ResponsePrepareProposal` and thus refers to the **current block**. Tendermint
+  will use it in the prepared proposal's header.
+* likewise, the transaction results from executing the prepared proposal are
+  provided in `ResponsePrepareProposal` and refer to the transactions in the
+  **current block**. Tendermint will use them to calculate the results hash
+  in the prepared proposal's header.
+* the consensus parameter updates and validator updates are also provided in
+  `ResponsePrepareProposal` and reflect the result of the prepared proposal's
+  execution. They come into force in height H+1 (as opposed to the H+2 rule
+  in next-block execution model).
+
+If the Application decides to keep the next-block execution model, it will not
+provide any data in `ResponsePrepareProposal`, other than an optionally modified
+transaction list.
+
+In the long term, the execution model will be set in a new boolean parameter
+*same_block* in `ConsensusParams`.
+It should **not** be changed once the blockchain has started unless the Application
```

Comment:
```suggestion
It **must not** be changed once the blockchain has started unless the Application
```

I feel like we can safely be prescriptive about this. The problem with saying "if you really know what you're doing" is people usually assume they do. I'd suggest we default to saying "don't do this" without qualification, unless we also want to explain somewhere what the criteria are for knowing when it's OK.